### DISCUSSION TITLE
Replacement of SDF dummy package

### DISCUSSION THREAD

**Discussion by Ciobrelis** [OpenModelica/OpenModelica#10041] [Q&A] [answered]
Hello, 

I wanted to use freely available AixLib library. However, it wouldn't load and gave me an error
 ![image](https://i.stack.imgur.com/7AYRw.png). 
After exchanging few messages on stackoverflow, it seems that the issue might be caused by missing SDF package. 

Although, there is one on OM, it is said to be a dummy giving the following: _"This is an open source dummy package of a commercial library. If you want it to be supported, you will need to add an implemention on our github repository: https://github.com/Modelica-OpenVendorSpecificLibraries."_ 

Sure enough, the package is empty: 
![image](https://user-images.githubusercontent.com/34867742/212083294-53ade1d6-b87a-41d8-9d8c-c7843c695bf4.png)

How to solve this issue? Is there a way to obtain SDF package? Preferably free, as that's the idea of using OM & freely available libraries. 

Versions (tested all):
AixLib 1.3.0 & 1.3.1
OM v1.19 & 1.20

**Comment by casella** [comment]
@Ciobrelis thanks for reporting! I checked the notifications when you load AixLib
```
[18] 19:14:12 Scripting Notification
Automatically loaded package NcDataReader2 2.5.0 due to uses annotation from AixLib.

[19] 19:14:12 Scripting Notification
Automatically loaded package Modelica 3.2.3 due to uses annotation from NcDataReader2.

[20] 19:14:12 Scripting Notification
Automatically loaded package Complex 4.0.0 due to uses annotation from Modelica.

[21] 19:14:12 Scripting Notification
Automatically loaded package ModelicaServices 4.0.0 due to uses annotation from Modelica.

[22] 19:14:12 Scripting Notification
[C:/Users/Francesco Casella/AppData/Roaming/.openmodelica/libraries/SDF master/package.mo: 9:42-9:258]: This is an open source dummy package of a commercial library. If you want it to be supported, you will need to add an implemention on our github repository: https://github.com/Modelica-OpenVendorSpecificLibraries.

[23] 19:14:12 Scripting Notification
Automatically loaded package SDF 0.0.0 due to uses annotation from AixLib.

[24] 19:14:12 Scripting Notification
Automatically loaded package Modelica_DeviceDrivers 2.0.0 due to uses annotation from AixLib.

[25] 19:14:12 Scripting Notification
Automatically loaded package ObsoleteModelica4 4.0.0 due to uses annotation from AixLib.
```
As I see it, there may be a problem with the dependencies. AixLib has
```modelica
 annotation (
  uses(
    NcDataReader2(version="2.5.0"),
    SDF(version="0.4.2"),
    Modelica_DeviceDrivers(version="2.0.0"),
    ObsoleteModelica4(version="4.0.0"),
    Modelica(version="4.0.0")),
  version="1.3.1",
...
```
so it uses NcDataReader2 v. 2.5.0, which _in the installed version_ has `uses(Modelica(version=3.2.3))`, which is of course in conflict with the uses(Modelica(version = "4.0.0")) of AixLib.

**Comment by casella** [comment]
The version management for NcDataReader2 is a bit messy. 

The original library is hosted at https://github.com/jraedler/ncDataReader2. It has no version and no uses annotation. The comment says the code is old and unmaintained, the last commit is from 2015.

The library we are installing is taken from https://github.com/modelica-3rdparty/netCDF-DataReader, which took over the development after 2015. There are no released versions, which makes the handling by the package manager a bit problematic.

All the commits from 2015 to 2019 were under the 2.5.0 version name. The very last commit changed the used version of Modelica from 3.2.3 to 4.0.0, and the version number from 2.5.0 to 2.5.1. This is quite questionable, as changing the used MSL from 3.2.3 to 4.0.0 means the new library is not backwards compatible in terms of dependencies (since MSL 3.2.3. is not backwards compatible with MSL 4.0.0), so this should not be a patch version but rather a new major version, i.e. 3.0.0, see Modelica/ModelicaSpecification#3209

**Comment by casella** [comment]
In any case, as long as AixLib uses NCDataReader2 version 2.5.0, and this uses MSL 3.2.3, this can't really work. I guess the cleanest solution is the following:
- Change the version number of the master/HEAD version of NCDataReader2 to 3.0.0, since it is not backwards compatible.
- Build two official 2.5.0 and 3.0.0 releases on https://github.com/modelica-3rdparty/netCDF-DataReader
- Change the uses annotation in AixLib 1.3.1 to use NCDataReader2 version 3.0.0

This requires action from the respective library developers. @overtaker, @tbeu, @FWuellhorst, what do you think?

**Comment by FWuellhorst** [comment]
@casella @Ciobrelis : 
The latest AixLib dev does not use NCDataReader any more, it's already in the `Obsolete` package and does not appear in the uses condition any more:
https://github.com/RWTH-EBC/AixLib/blob/development/AixLib/package.mo

In any case @Ciobrelis : Please check out the branch https://github.com/RWTH-EBC/AixLib/tree/openModelicaProject, which contains changes and fixes to increase the compatibility to OM. Sadly, the review process of the PR takes some time. 

Regarding SDF @casella :
Do I get it right that we would have to implement an own open-source version of the SDF library to be able to use it in OM? So far, I just loaded the commercial version.

**Reply by Ciobrelis** [reply]
@FWuellhorst I would suggest also changing `Modelica_DeviceDrivers` version from `2.0.0` to `2.1.1`, as otherwise it prompts warning errors, when I check examples (e.g. `AixLib.ThermalZones.HighOrder.Examples.OFD_1Jan` or `AixLib.ThermalZones.HighOrder.Examples.RoomGFOw2_DayNightMode`) among others. 
![image](https://user-images.githubusercontent.com/34867742/212376296-addb88a3-87fc-4558-8122-dca3edb2b13e.png)

However, these warning disappear if I use `2.1.1`.

**Reply by FWuellhorst** [reply]
Thanks for noticing! These warnings always annoyed me, didn't know 2.1.1 solves this issue.

**Comment by casella** [comment, marked as answer]
@FWuellhorst thanks for the quick reaction. Library management is a bit of a nightmare 😅 

In fact, I was wondering why the [library testing report](https://libraries.openmodelica.org/branches/master/AixLib_latest/AixLib_latest.html) didn't have this problem. The answer is: because we are only testing the latest version on the development branch. However, the versions that are installed by the package manager are different: they are the versions that are marked as "Released" on GitHub, see https://github.com/RWTH-EBC/AixLib/releases. This includes the 1.3.1 pre-release, which still contains the uses NcDataReader2 annotation. I installed 1.3.0, but it suffers from the same issue when I try to open it. Unfortunately, we are not testing any of them in our CI system as of today.

So, I would suggest the following course of action
- We declare the AixLib library as "experimental", not "supported", until these issues are sorted out.
- We add the test of the latest release (or pre-release) to our CI, besides the latest development version.
- I would recommend that you make a new 1.3.1 pre-release that incorporate all improvements for OMC compatibility, including the updated uses annotation
- Once you've done that, we can declare 1.3.1 "partially supported", since about 80% of the runnable models in the latest development version are simulated successfully. Some of the remaining 20% of the models are not running because they are not fully compliant to the Modelica specification, see e.g. [this one](https://libraries.openmodelica.org/branches/master/AixLib_latest/files/AixLib_latest_AixLib.Media.Refrigerants.Examples.RefrigerantProperties.err), which is using a class from a partial package, which is forbidden by the rules in [Section 5.3.2](https://specification.modelica.org/maint/3.5/scoping-name-lookup-and-flattening.html#composite-name-lookup). Others may be broken because of OMC bugs, in which case please open tickets about them [here](https://github.com/OpenModelica/OpenModelica/issues/new/choose).
- Next versions of the library that have a higher success ratio could be declared as fully supported

@FWuellhorst are you OK with this plan?

**Reply by FWuellhorst** [reply]
This plan looks good!
Regarding the "partially supported": We plan to get our branch for fixing the OM bugs merged this month. After this, all remaining bugs are the ones already raised to OM. Coverage rate should be much higher. Thus, I hope we get the "fully supported" status rather quickly.

**Comment by Ciobrelis** [comment]
Okay, so I found out, that the AixLib library available on Git differs from the one that is being installed on OM via `Manage Libraries ->Install Library` although the newest version is chosen. 
In both cases, it says the version is 1.3.1, but after checking package.mo I see the following (right - from Git). 
![image](https://user-images.githubusercontent.com/34867742/212357353-f3012a74-c66e-4fbc-87a3-efee67508b60.png)

If I use the version from Git it loads without any problems. Otherwise, one has to implement changes suggested by @casella

**Comment by casella** [comment]
@Ciobrelis, the way we configured the package manager currently makes all versions appearing here https://github.com/RWTH-EBC/AixLib/releases from 0.9.1 available for download and marked as "supported". It does it automatically also as new versions are released. This is of course wrong, because if they don't work, they shouldn't be marked as that, I'll fix that right away. Version 1.3.1 is the one that you tagged as pre-release (b341e69), which is also still not good because of the old uses annotations.

So, I'd suggest that you remove the current 1.3.1 pre-release and make another one that includes all those fixes, so that at least 1.3.1 works out of the box in OpenModelica.

I see another problem, the current 1.3.1 is marked as pre-release on Github, but the OMC package manager sees that as a final release, so it doesn't show up as such (e.g. as 1.3.1 pre-release) in the list of versions that you can install. We have means to identify pre-releases, beta releases or post-release builds by adding prefixes and postifixes to the version tag (as specified in [semver](https://semver.org/spec/v2.0.0-rc.2.html)), but we need to do that manually. I can do that for this version if you want, then we can agree on what to do in the future, if you are interested.

**Reply by Ciobrelis** [reply]
I added library manually. As of now it works and I am also aware of the situation, so all good from my side. Thank you! 

Another aspect that might be related to `Libraries Management`. Not sure whether it is a problem or not. Let's take an example with `Modelica_DeviceDrivers` library as it prompts warnings with AixLib. On my OM I had v2.0.0 but I saw that there is v2.1.1 on Git. If I choose `Upgrade Installed Libraries`, the version remains 2.0.0 without upgrading it, although I can find and install v2.1.1 via `Manage Libraries ->Install Library`.

**Comment by casella** [comment]
I have updated the package manager server setups. In a couple of days, when all the periodic jobs have ran, you should see:
- two separate test reports [here](https://libraries.openmodelica.org/branches/master/): `AixLib` with the latest release version (currently the fixed 1.3.1 pre-release), and _AixLib_latest_ with the latest commit on the development branch
- only version 1.3.1 showing up in the OMEdit package manager drop-down menu, if you leave the default checks on support and fullSupport - you may need to update the library index first

**Comment by casella** [comment]
@FWuellhorst regarding SDF, I understand there is a Dymola library to handle that, I don't know the details, but I guess it is using some Dymola internals. If you are interested in an open-source implementation, please let me know, so we can coordinate with the Modelica Standard Library development.

**Reply by Ciobrelis** [reply]
Do you have an example how SDF should look like on Modelica?

**Reply by FWuellhorst** [reply]
@casella : 
We would be really interested in an open-source implementation.
We heavily use performance maps for simulation of complex devices such as the vapor compression cycle in heat pumps in bigger systems over longer time periods.
The 2D approach was fine for on/off devices. Now that modulation is state of the art, we need higher dimensions for these maps. 
However, MSL decided against it some years ago: https://github.com/modelica/ModelicaStandardLibrary/issues/1516
Maybe you can change their opinion?

@Ciobrelis 
As far as I can see, the equations are purely external C. I know of open-source python libraries, so external Python should work rather quickly. But if someone knows C, implementing these external functions should not be too much work. There is already one C-Library for it:https://github.com/ScientificDataFormat/NDTable/tree/develop

**Comment by casella** [comment]
I had a quick look at modelica/ModelicaStandardLibrary#1516, as I understand the controversy was on the library not being open-source enough.

The problem with these proposals is often that they start from some existing implementation, and then try to fit it in the framework, which is sometimes not successful.

I would suggest that you open a fresh ticket on the MSL issue tracker where you state what are your requirements. I can then try to see if there is some shared interest, and if there is, we can also try to get some paid development for that, since this feature seems to be quite useful in a number of applications.

**Reply by beutlich** [reply]
There still is https://github.com/modelica/ModelicaStandardLibrary/issues/1153 with the one open task: Support n-dimensional tables

**Reply by FWuellhorst** [reply]
@beutlich Thanks for pointing that out. Is this development planned in the near future? If not, would it help to comment on that issue explicitly stating the need?

**Reply by casella** [reply]
I'm not aware of any current plan to do that. Commenting may help for sure, try to be as precise as possible about your requirements.

**Comment by FWuellhorst** [comment]
@casella : I just released the latest AixLib version 1.3.2 which increases the OM compatibility (at least locally) drastically. The remaining issues are either raised to OM or being resolved in other libraries. 
Further, we are in the final steps to include an OM image into our CI to prevent future version from bad compatibility to OM. 

Hope that the changes lead to a new status in OM :)

**Reply by casella** [reply]
See [regression report](https://libraries.openmodelica.org/branches/history/master/2023-02-09%2000:52:40..2023-02-09%2013:50:21.html).

**Reply by FWuellhorst** [reply]
Thanks! 
Only the models under `AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140` do simulate under Windows version ` v1.19.2 (64-bit)`. 

I don't find why in the [report ](https://libraries.openmodelica.org/branches/master/AixLib/files/AixLib_AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140.FourElements.TestCase600.err). 
Using the latest `1.21.0-dev` on Windows, the models fail as well:

```
.../OpenModelica/OMEdit/AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140.FourElements.TestCase600/TestCase600.exe -port=52398 -logFormat=xmltcp -override=startTime=0,stopTime=3.1536e+07,stepSize=3600,tolerance=1e-06,solver=dassl,outputFormat=mat,variableFilter=.* -r=.../OpenModelica/OMEdit/AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140.FourElements.TestCase600/TestCase600_res.mat -w -lv=LOG_STATS -inputPath=.../OpenModelica/OMEdit/AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140.FourElements.TestCase600 -outputPath=.../OpenModelica/OMEdit/AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140.FourElements.TestCase600
... loading "Table" from "D:\04_git\AixLib\AixLib\Resources\WeatherData\WeatherData_Ashrae140_LOM.mat" 
... loading "Table" from "D:\04_git\AixLib\AixLib\Resources\WeatherData\Weatherdata_ASHARE140.mat" 
The initialization finished successfully without homotopy method.
The arguments u_min and u_max provided in the function call     solveOneNonlinearEquation(f,u_min,u_max) do not bracket the root of the single non-linear equation 0=f(u):   u_min  = 200   u_max  = 6000   fa = f(u_min) = 235966   fb = f(u_max) = 7.55392e+06 fa and fb must have opposite sign which is not the case
Process crashed
```

The error ` The arguments u_min and u_max provided in the function call     solveOneNonlinearEquation(f,u_min,u_max) do not bracket the root of the single non-linear equation 0=f(u):   u_min  = 200   u_max  = 6000   fa = f(u_min) = 235966   fb = f(u_max) = 7.55392e+06 fa and fb must have opposite sign which is not the case` does appear in the log under 1.19 as well, but does not lead to a crash.

I can solve the issue in AixLib, just as a note that some outputs are missing in the regression report :)

**Comment by casella** [comment]
@FWuellhorst I'm a bit lost. Models under `AixLib.ThermalZones.ReducedOrder.Validation.ASHRAE140` seems to work fine with the master version of the compiler, see [report](https://libraries.openmodelica.org/branches/master/AixLib/AixLib.html). Only PartialCase fails, which is farly obvious, it probably shouldn't have an experimente annotation since it cannot be simulated. 

To be more precise, they work in [v1.19](https://libraries.openmodelica.org/branches/v1.19/AixLib/AixLib.html), fail in [v1.20](https://libraries.openmodelica.org/branches/v1.20/AixLib/AixLib.html), and work again on [master](https://libraries.openmodelica.org/branches/master/AixLib/AixLib.html) (v1.21-dev).

**Reply by FWuellhorst** [reply]
Ah ok, I thought that this red color indicates some sort of error:
![image](https://user-images.githubusercontent.com/22442514/218513638-d30e6116-640f-4fa7-a467-eace60fe3499.png)

Then it's all fine! We will try to fix the failing models in the upcoming weeks! Thanks for the help and the nice regression reports!

**Reply by casella** [reply]
> Ah ok, I thought that this red color indicates some sort of error:

It does, that's the 1.20 report, and there the simulations are indeed failing. Since the error is in the "Simulate" column, you can then check on the "sim" link to see the simulation log and get some feedback about the cause of failure. Failures in previous stages of code generation (last five columns) can be investigated by checking the simulation log, using the link with the model name. If that is successful, clicking on that just shows you the completed simulation log and the command line used to start the simulation, but not the simulation log.

> We will try to fix the failing models in the upcoming weeks

No real need to do that for those models that work with the master version. In a couple of months or so we should release 1.21.0, and they will just work there. Unless you need an official version that handles them right now, in which case we could try to figure out which are the commits that fixed the problem on master and cherry pick them for a maintenance release 1.20.1, but that would take some work.

**Reply by FWuellhorst** [reply]
Ok, perfect, thanks for the explanation! That makes sense.
